<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        let p1 = function () {
          return new Promise((resolve, reject) => {
                setTimeout(function () {
                    console.log('p1');
                    resolve('p1xxxxx');// resolve 输出异步结束的 成功结果
                }, 1000)
            })
        }

        let p2 = function () {
            return new Promise((resolve, reject) => {
                setTimeout(function () {
                    console.log('p2');
                    resolve('p2xxxxxxxx');
                }, 2000)
            })
        }


        // async await 
        // 在普通函数 前 加关键字 async 就是async函数
        // awiait 
        // 1: 只能在async 函数内部使用
        // await p1()
        // 2: await 后哦面必须赋值为promise 实例 或者 async函数,
        // ------> await 后promise实例产生 1：函数执行；2 new Promise 3:Promise api返回值 then  all...
        // 3: await 的返回值，相当于 then 的回参赋值 (关键看await 后promise 是如何产生的)
        // 3.1  resolve  
        // 3.2  then 中 return 

        async function asy(){
            let res = await p1();
            console.log(res,'res');
            let res1 = await new Promise((resolve,reject)=>{
                resolve("new promise")
            })
            console.log(res1);
            let res2 = await p1().then(data=>{
                return data+'then returm'
            })
            console.log(res2,'res2');
        }
        // asy()


        // 优点
        async function asy1(){
            // 1:异步问题，同步化
            let res = await p2();
            console.log(res);
            // 等上一个 await 执行完毕后 在执行下面的await
            let res1 = await p1();
            console.log(res1);
            // await 返回值；就是异步执行结束，输出的结果。
        }
        // asy1()

        // 捕获错误
        async function asy2(){
            try{ // 正确代码
                // let res = await p2();
                // console.log(res);
                // let re2 = await p1();
                // console.log(re2);
                return 'asy2'
            }catch(err){ // 捕获错误
                console.log('这里可以捕获到 try 中执行的错误',err);
            }
        }
        
        // async 中return 问题
        async function asy3(){
            let res = await asy2();  
            console.log(res,'asy3 res');
            return await p1();
        }
        // async 函数执行 返回值为promise 
        let promise = asy3() // promise
        // return 传值到 then 的回参中
        promise.then(res=>console.log(res));// 1


        // async await 是异步程序。（慢 await）
        async function ayn4(){
            console.log(asy3());
            // await 接受 asy3 中 return 的await 返回值
            let res = await asy3();
            console.log(res,'接受await返回值');
        }
        ayn4()

        // await 异步 规律
        // 1: res = await xxx();
        // 1.1 xx 是否是async 函数
        // 不是：返回值一定是 promise
        // 是 
        // 如果返回值 js 数据；await返回数据 res = 数据
        // 如果返回是另一个(await jjj()) 那么 await返回就是jjj()的返回值（基本上都是promise）
        
    </script>
</body>

</html>